### ISSUE TITLE
Desktop Client

### ISSUE THREAD

**Issue by daniellim2000** [codecombat/codecombat#3504]
It would be good to have a desktop client that stores the stuff in the user's computer instead of having to redownload it every time the user goes to the site, as it is very inefficient for people(like me) who are having terrible network conditions.

**Comment by Zeldaretter** [comment]
That is a great idea. Especially if you use it in schools the network connection is often shared with many devices and so the speed gets terrible over time.

**Comment by daniellim2000** [comment]
Yes exactly I'm staying at hostel and I almost cannot access the levels. T.T
Would be nice if we could make a team to do this together.

**Comment by nwinter** [comment]
Sounds like a great project to try out. What's the easiest way of doing this–a Chrome app?

**Comment by Zirientis** [comment]
Possibly, or it may be an option to use the Qt framework (it includes a way to embed HTML). The only interesting part I see then would be to change the code so that data is preferentially loaded from the user's computer rather than the site. Qt would be available under the LGPL or a commercial license. This way, we could also target people who don't use Chrome.

**Comment by basicer** [comment]
Hmm, I would think the easiest would be to use electron. http://electron.atom.io/  It's based on node, which we use, and also has a top notch browser built into for rendering things.

**Comment by nwinter** [comment]
Oh yeah, totally! Someone want to get a test Electron + CodeCombat project going?

**Comment by daniellim2000** [comment]
Sure no problem

**Comment by UltCombo** [comment]
Electron is a great solution for desktop apps (Node.js + Chromium FTW!), but I'm not sure how well-suited it is for the CodeCombat case. I mean, sure, you can bundle the levels and Thangs with the executable so you can download it all upfront, but these things get outdated fast and you will likely need to fetch newer versions from the server. In any case, you will still need an working internet connection in order to login, submit level solutions, make item purchases etc. I believe this all diminishes the usefulness of a CodeCombat desktop client.

If you just want smarter caching and the ability to download all resources upfront for a better experience on shaky connections later on (I guess these are the motivating reasons behind this thread?), then you don't need a desktop client for that—I would suggest looking into [Service Workers](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API/Using_Service_Workers), more specifically its cache API. And if you want some good ol' abstraction to get things done faster, take a look at the [Service Worker Precache module](https://github.com/GoogleChrome/sw-precache), as well as other [Service Worker Libraries](https://developers.google.com/web/tools/service-worker-libraries/?hl=en).

**Comment by nwinter** [comment]
Oh man, service workers look cool; we should try those out out for better asset caching. Separately, having a simpler Electron wrapper might be good because we could add a different per-seat instead of per-student pricing model, which some libraries and other types of customers kind of want.

**Comment by UltCombo** [comment]
As we have talked in Slack previously, using the Cache API alone is not bullet-proof as browsers may empty the cache if the user does not visit frequently or if the disk space is running low.

A complementary solution would be use the [Storage API](https://storage.spec.whatwg.org/#introduction)'s `navigator.storage.persist()` method when it is available, which Blink (Chrome) is [willing to ship](https://groups.google.com/a/chromium.org/forum/m/#!msg/blink-dev/5Sihi1iAXYc/SISKuRppAwA):

> **Summary**
> Allow origins to opt out of the browser's storage-eviction logic that is run when the user's storage space is running low. For privacy reasons, the data is still cleared if the user invokes “clear browsing data”.
>
> **Motivation**
> The web platform, as specified, doesn't require browsers to guarantee the lifetime of web storage (e.g. IndexedDB, WebSQL, Service Workers). In practice browsers implement storage on a "best-effort" basis: when an end-user's storage space runs low, they start deleting web storage. This means that (for example) there is no way for a music service to guarantee that the songs you saved for offline won't be deleted by the browser behind your back. To get around this shortcoming, developers either go native or use chrome apps and extensions with the "unlimitedStorage" permission
